### PR TITLE
[exporter/loadbalancing] Improve the performance when merging traces belonging to the same backend

### PR DESCRIPTION
**Pull request by ghost**
**Description:** no need to reimplement that in an extremely allocation-inefficient fashion.

I'm actually not sure why mergeTraces() and mergeMetrics() need to exist in the first place; all the other exporters coupled with the batch processor work just fine, not sure why loadbalancing would be special. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30141 seems to imply they were implemented to improve performance, but I don't really understand why batch processor would not have been sufficient for that improvement originally.

benchmarks before:
```
	goos: darwin
	goarch: arm64
	pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter
	BenchmarkMergeTraces_X100-8     	   50214	     23507 ns/op
	BenchmarkMergeTraces_X500-8     	   10000	    113952 ns/op
	BenchmarkMergeTraces_X1000-8    	    5208	    226062 ns/op
	BenchmarkMergeMetrics_X100-8    	   64933	     18540 ns/op
	BenchmarkMergeMetrics_X500-8    	   12885	     91418 ns/op
	BenchmarkMergeMetrics_X1000-8   	    6590	    184584 ns/op
	PASS
	ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter	9.783s
```
and after:
```
	goos: darwin
	goarch: arm64
	pkg: github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter
	BenchmarkMergeTraces_X100-8     	295886529	         3.836 ns/op
	BenchmarkMergeTraces_X500-8     	309865370	         3.833 ns/op
	BenchmarkMergeTraces_X1000-8    	310739948	         3.800 ns/op
	BenchmarkMergeMetrics_X100-8    	315567813	         3.841 ns/op
	BenchmarkMergeMetrics_X500-8    	310341650	         3.849 ns/op
	BenchmarkMergeMetrics_X1000-8   	314292003	         3.830 ns/op
	PASS
	ok  	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/loadbalancingexporter	10.733s
```
**Link to tracking Issue:** n/a
**Testing:** unit tests pass & cpu time for our collectors using loadbalancingexporter (12 replicas, total of 25k-40k spans/sec) went from 800ms-1400ms/sec down to <40msec/sec.
**Documentation:** none

### CODE REVIEW

**Comment by lauri-paypay** [comment]
@jpkrohling @bogdandrutu Ping; could I get a review on this? We've been running this internally without problems since I created this PR.

**Comment by lauri-paypay** [comment]
@dmitryax sorry for mentioning you directly, but since you seem to have been active in maintaining loadbalancingexporter, could I ask you to take a look at this PR?

**Comment by jpkrohling** [comment]
@lauri-paypay, thank you for the PR. The problem that PR was solving is optimizing the number of requests made by the load balancing exporter to the next layer of collectors. By merging traces per endpoint, we are making traces that are meant to the same destination to be sent on a single payload. 

Your benchmark here is good, but it focuses only on a specific function call. Would it be possible to also compare the number of requests (or calls to `consumeMetric`)? Or even better, perhaps we could also report the number of entries on the payload as part of one of the available metrics?

**Comment by lauri-paypay** [comment]
>Your benchmark here is good, but it focuses only on a specific function call. Would it be possible to also compare the number of requests (or calls to consumeMetric)? Or even better, perhaps we could also report the number of entries on the payload as part of one of the available metrics?

I'm sure it's possible, but I don't think that should be done as part of this PR, because this PR changes only the function that I did test.

File: .chloggen/loadbalancingexporter_perf.yaml
Comment (jpkrohling) [inline .chloggen/loadbalancingexporter_perf.yaml:10]:
```suggestion
note: Improve the performance when merging traces belonging to the same backend 
```

**Review by jpkrohling — COMMENTED** [review]
> PR does not change behavior

That's true, we have tests asserting this. Thank you! 

I still feel like we are missing something, but I'm OK giving this a go, as we do have test coverage for the scenario I have in mind.

The only request I have: can we make the changelog entry a bit easier for users to find out what's happening. In case of problems, they should be able to come back to this PR and let us know they are having problems.

**Comment by lauri-paypay** [comment]
> I still feel like we are missing something, but I'm OK giving this a go, as we do have test coverage for the scenario I have in mind.
> 
> The only request I have: can we make the changelog entry a bit easier for users to find out what's happening. In case of problems, they should be able to come back to this PR and let us know they are having problems.

thanks. I committed your suggestion to address that.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
We are currently blocked by a new vulnerability in Go, which is causing our vulnerability check workflow to fail. We'll merge this PR once that is fixed.

**Pull request merged**
